File: src/graph.py
# src/graph.py

from PyQt5 import QtWidgets, QtCore, QtGui
import pyqtgraph as pg
import numpy as np
from scipy.signal import savgol_filter

WINDOW_LENGTH = 13
POLYORDER = 2

class GazeTrackingPlotter(QtWidgets.QWidget):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("Real-Time Gaze Tracking")
        self.layout = QtWidgets.QVBoxLayout()
        self.setLayout(self.layout)

        # X and Y Line Plots
        self.plot_widget_x = pg.PlotWidget(title="X Coordinate Over Time")
        self.plot_widget_y = pg.PlotWidget(title="Y Coordinate Over Time")
        self.layout.addWidget(self.plot_widget_x)
        self.layout.addWidget(self.plot_widget_y)

        # 2D Scatter Plot
        self.plot_widget_2d = pg.PlotWidget(title="2D Gaze Points Over Time")
        self.layout.addWidget(self.plot_widget_2d)

        # 3D Scatter Plot (using pyqtgraph's GLViewWidget)
        self.plot_widget_3d = pg.opengl.GLViewWidget()
        self.scatter_3d = pg.opengl.GLScatterPlotItem()
        self.plot_widget_3d.addItem(self.scatter_3d)
        self.layout.addWidget(self.plot_widget_3d)

    def update_plots(self, time_data, x_data, y_data):
        # Apply smoothing
        x_data_smoothed = savgol_filter(x_data, WINDOW_LENGTH, POLYORDER) if len(x_data) >= WINDOW_LENGTH else x_data
        y_data_smoothed = savgol_filter(y_data, WINDOW_LENGTH, POLYORDER) if len(y_data) >= WINDOW_LENGTH else y_data

        # Update Line Plots
        self.plot_widget_x.clear()
        self.plot_widget_y.clear()
        self.plot_widget_x.plot(time_data, x_data_smoothed, pen='r')
        self.plot_widget_y.plot(time_data, y_data_smoothed, pen='b')

        # Update 2D Scatter Plot
        self.plot_widget_2d.clear()
        scatter_2d = pg.ScatterPlotItem(x=x_data, y=y_data, pen=None, brush=pg.intColor(200, 255))
        self.plot_widget_2d.addItem(scatter_2d)

        # Update 3D Scatter Plot
        colors = np.array([[1, 0, 0, 1]] * len(time_data))  # Red points
        pos = np.column_stack((x_data, y_data, time_data))
        self.scatter_3d.setData(pos=pos, color=colors, size=5)


def plot_gaze_tracking(time_data, x_data, y_data, ax_x=None, ax_y=None, scatter_ax=None, scatter3d_ax=None):
    app = QtWidgets.QApplication([])
    plotter = GazeTrackingPlotter()
    plotter.update_plots(time_data, x_data, y_data)
    plotter.show()
    app.exec_()


def plot_final_graphs(time_data, x_data, y_data, save_path='final_comprehensive_plots.png'):
    app = QtWidgets.QApplication([])

    # Create main widget to store final plots
    final_plotter = GazeTrackingPlotter()
    final_plotter.update_plots(time_data, x_data, y_data)

    # Capture and save the widget as an image
    screenshot = final_plotter.grab()
    screenshot.save(save_path, 'PNG')

    print(f"Final comprehensive plots saved as {save_path}.")
    final_plotter.show()
    app.exec_()

File: src/graph.js
const Plotly = require('plotly.js-dist');
const fs = require('fs');
const smooth = require('smooth');

// Smoothing parameters
const WINDOW_LENGTH = 13;
const POLYORDER = 2;

class GazeTrackingPlotter {
    constructor() {
        this.initPlots();
    }

    initPlots() {
        // Initialize X and Y line plots
        this.plotDivX = document.createElement('div');
        this.plotDivY = document.createElement('div');
        document.body.appendChild(this.plotDivX);
        document.body.appendChild(this.plotDivY);

        // Initialize 2D scatter plot
        this.plotDiv2D = document.createElement('div');
        document.body.appendChild(this.plotDiv2D);

        // Initialize 3D scatter plot
        this.plotDiv3D = document.createElement('div');
        document.body.appendChild(this.plotDiv3D);
    }

    updatePlots(timeData, xData, yData) {
        // Apply smoothing (savgol-like filter)
        const xDataSmoothed = smooth(xData, { method: 'savgol', window: WINDOW_LENGTH, polynomial: POLYORDER });
        const yDataSmoothed = smooth(yData, { method: 'savgol', window: WINDOW_LENGTH, polynomial: POLYORDER });

        // Plot X Coordinate Over Time
        Plotly.newPlot(this.plotDivX, [{
            x: timeData,
            y: xDataSmoothed,
            mode: 'lines',
            name: 'X Coordinate',
            line: { color: 'red' }
        }], { title: 'X Coordinate Over Time' });

        // Plot Y Coordinate Over Time
        Plotly.newPlot(this.plotDivY, [{
            x: timeData,
            y: yDataSmoothed,
            mode: 'lines',
            name: 'Y Coordinate',
            line: { color: 'blue' }
        }], { title: 'Y Coordinate Over Time' });

        // Plot 2D Gaze Points Over Time
        Plotly.newPlot(this.plotDiv2D, [{
            x: xData,
            y: yData,
            mode: 'markers',
            marker: { color: 'rgba(255, 0, 0, 0.8)' }
        }], { title: '2D Gaze Points Over Time', xaxis: { title: 'X' }, yaxis: { title: 'Y' } });

        // Plot 3D Gaze Points
        Plotly.newPlot(this.plotDiv3D, [{
            x: xData,
            y: yData,
            z: timeData,
            mode: 'markers',
            marker: { size: 5, color: 'rgba(255, 0, 0, 0.8)' },
            type: 'scatter3d'
        }], { title: '3D Gaze Points' });
    }
}

function plotGazeTracking(timeData, xData, yData) {
    const plotter = new GazeTrackingPlotter();
    plotter.updatePlots(timeData, xData, yData);
}

function plotFinalGraphs(timeData, xData, yData, savePath = 'final_comprehensive_plots.html') {
    const plotter = new GazeTrackingPlotter();
    plotter.updatePlots(timeData, xData, yData);

    // Save final plot as HTML for viewing in a browser
    const htmlContent = `
        <html>
            <head>
                <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
            </head>
            <body>
                ${plotter.plotDivX.outerHTML}
                ${plotter.plotDivY.outerHTML}
                ${plotter.plotDiv2D.outerHTML}
                ${plotter.plotDiv3D.outerHTML}
            </body>
        </html>`;
    fs.writeFileSync(savePath, htmlContent);
    console.log(`Final comprehensive plots saved as ${savePath}.`);
}

// Example usage:
const timeData = [0, 1, 2, 3, 4];  // Example time data
const xData = [100, 150, 200, 250, 300];  // Example X coordinates
const yData = [200, 250, 300, 350, 400];  // Example Y coordinates

plotGazeTracking(timeData, xData, yData);
plotFinalGraphs(timeData, xData, yData);
